File: scripts/lint-no-hebrew-in-code.py
#!/usr/bin/env python3
"""
Block hardcoded Hebrew strings in source code.

Scans TS + Svelte files for Hebrew characters (U+0590..U+05FF) that appear
inside string literals (single/double/backtick quotes) or Svelte template
text. Hebrew inside line comments (`//`) and block comments (`/* */`) is
allowed — those are developer notes.

Scope:
  - packages/frontend/
  - packages/core/
  - packages/backend/

Allowlist (paths where Hebrew IS allowed in strings):
  - packages/core/src/i18n/catalogs/*  (source of truth)
  - packages/core/src/voice/*-prompt.ts  (LLM prompts)
  - **/*.test.ts, **/tests/**, **/fixtures/**  (test data)

Exit 0 = clean. Exit 1 = violations found.
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

SCAN_DIRS = [
    "packages/frontend",
    "packages/core",
    "packages/backend",
]

ALLOW_PATTERNS = [
    re.compile(r"packages/core/src/i18n/catalogs/"),
    re.compile(r"/voice/.*-prompt\.ts$"),
    re.compile(r"packages/backend/src/prompts/"),  # LLM system prompts (slice 14)
    re.compile(r"\.test\.ts$"),
    re.compile(r"/tests/"),
    re.compile(r"/fixtures/"),
    re.compile(r"/node_modules/"),
    re.compile(r"/build/"),
    re.compile(r"/\.svelte-kit/"),
    re.compile(r"/dist/"),
]

HEBREW_RE = re.compile(r"[\u0590-\u05FF]")


def is_allowed(path: Path) -> bool:
    rel = str(path.relative_to(REPO_ROOT))
    return any(p.search(rel) for p in ALLOW_PATTERNS)


def strip_jsdoc_blocks(text: str) -> str:
    """
    Pre-pass: blank out all `/** ... */` and `/* ... */` blocks.
    Done before the main state machine so regex literals containing
    quotes (`/.../`) don't confuse comment detection.

    Preserves line breaks + column counts.
    """
    out = list(text)
    i = 0
    n = len(text)
    while i < n - 1:
        if text[i] == "/" and text[i + 1] == "*":
            end = text.find("*/", i + 2)
            if end == -1:
                end = n
            else:
                end += 2
            for j in range(i, end):
                if out[j] != "\n":
                    out[j] = " "
            i = end
        else:
            i += 1
    return "".join(out)


def strip_all_comments(text: str) -> str:
    """
    Walk the full file as a single character stream with a state machine.
    Replace every comment character with a space, preserving line breaks
    and column counts. Strings are preserved as-is.

    States: code | line_comment | str_dq | str_sq | str_bt
    (Block comments are handled in strip_jsdoc_blocks before this.)
    """
    text = strip_jsdoc_blocks(text)
    out: list[str] = []
    state = "code"
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        nxt = text[i + 1] if i + 1 < n else ""
        if state == "code":
            if ch == "/" and nxt == "/":
                out.append("  ")
                state = "line_comment"
                i += 2
                continue
            if ch == '"':
                state = "str_dq"
            elif ch == "'":
                state = "str_sq"
            elif ch == "`":
                state = "str_bt"
            out.append(ch)
            i += 1
            continue
        if state == "line_comment":
            if ch == "\n":
                out.append("\n")
                state = "code"
            else:
                out.append(" ")
            i += 1
            continue
        # In a string literal: handle escapes, preserve text.
        if state in ("str_dq", "str_sq", "str_bt"):
            quote = {"str_dq": '"', "str_sq": "'", "str_bt": "`"}[state]
            if ch == "\\" and i + 1 < n:
                out.append(ch)
                out.append(text[i + 1])
                i += 2
                continue
            if ch == quote:
                out.append(ch)
                state = "code"
                i += 1
                continue
            out.append(ch)
            i += 1
            continue
    return "".join(out)


def scan_file(path: Path) -> list[tuple[int, str]]:
    """Return list of (line_no, original_line) where Hebrew appears outside comments."""
    try:
        text = path.read_text(encoding="utf-8")
    except (UnicodeDecodeError, OSError):
        return []
    code_only = strip_all_comments(text)
    violations: list[tuple[int, str]] = []
    original_lines = text.splitlines()
    code_lines = code_only.splitlines()
    for n, (raw, stripped) in enumerate(zip(original_lines, code_lines), start=1):
        if HEBREW_RE.search(stripped):
            violations.append((n, raw.rstrip()))
    return violations


def main() -> int:
    total_violations = 0
    bad_files: list[tuple[Path, list[tuple[int, str]]]] = []

    for d in SCAN_DIRS:
        root = REPO_ROOT / d
        if not root.exists():
            continue
        for ext in ("*.ts", "*.svelte"):
            for path in root.rglob(ext):
                if is_allowed(path):
                    continue
                v = scan_file(path)
                if v:
                    bad_files.append((path, v))
                    total_violations += len(v)

    if bad_files:
        print("Hebrew strings found in code (must use t('key') instead):\n", file=sys.stderr)
        for path, violations in bad_files:
            rel = path.relative_to(REPO_ROOT)
            print(f"  {rel}", file=sys.stderr)
            for n, line in violations:
                print(f"    {n}: {line}", file=sys.stderr)
            print("", file=sys.stderr)
        print(f"✗ {len(bad_files)} file(s), {total_violations} occurrence(s).", file=sys.stderr)
        print("", file=sys.stderr)
        print("Fix: add the string to packages/core/src/i18n/catalogs/{he,en}.ts,", file=sys.stderr)
        print("  add the key to packages/core/src/i18n/keys.ts,", file=sys.stderr)
        print("  and use t('your.key') in the component.", file=sys.stderr)
        return 1

    print("✓ No hardcoded Hebrew in code.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/lint-no-hebrew-in-code.mjs
#!/usr/bin/env node
// @ts-check
/**
 * Block hardcoded Hebrew strings in source code.
 *
 * Scans TS + Svelte files for Hebrew characters (U+0590..U+05FF) that appear
 * inside string literals (single/double/backtick quotes) or Svelte template
 * text. Hebrew inside line comments (`//`) and block comments (`/* *​/`) is
 * allowed — those are developer notes.
 *
 * Scope:
 *   - packages/frontend/
 *   - packages/core/
 *   - packages/backend/
 *
 * Allowlist (paths where Hebrew IS allowed in strings):
 *   - packages/core/src/i18n/catalogs/*  (source of truth)
 *   - packages/core/src/voice/*-prompt.ts  (LLM prompts)
 *   - **​/*.test.ts, **​/tests/**, **​/fixtures/**  (test data)
 *
 * Exit 0 = clean. Exit 1 = violations found.
 *
 * Pure Node stdlib (fs/path) — no dependencies, no build step. Runs via the
 * `.sh` wrapper which prefers this over the legacy Python implementation.
 */

import { readdirSync, readFileSync, statSync } from "node:fs"
import { dirname, join, relative, resolve } from "node:path"
import { fileURLToPath } from "node:url"

const REPO_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..")

const SCAN_DIRS = ["packages/frontend", "packages/core", "packages/backend"]

/** Paths (matched against the POSIX-style relative path) where Hebrew IS allowed. */
const ALLOW_PATTERNS = [
  /packages\/core\/src\/i18n\/catalogs\//,
  /\/voice\/.*-prompt\.ts$/,
  /packages\/backend\/src\/prompts\//, // LLM system prompts (slice 14)
  /\.test\.ts$/,
  /\/tests\//,
  /\/fixtures\//,
  /\/node_modules\//,
  /\/build\//,
  /\/\.svelte-kit\//,
  /\/dist\//,
]

const HEBREW_RE = /[\u0590-\u05FF]/

/**
 * @param {string} absPath
 * @returns {boolean}
 */
function isAllowed(absPath) {
  // Normalise to POSIX separators so the allow patterns match on every OS.
  const rel = relative(REPO_ROOT, absPath).split("\\").join("/")
  return ALLOW_PATTERNS.some((p) => p.test(rel))
}

/**
 * Pre-pass: blank out all `/​** ... *​/`, `/​* ... *​/`, and `<!-- ... -->` blocks.
 * Done before the main state machine so regex literals containing quotes
 * (`/.../`) don't confuse comment detection. Preserves line breaks.
 *
 * @param {string} text
 * @returns {string}
 */
function stripJsdocBlocks(text) {
  const out = [...text]
  const n = text.length
  let i = 0
  while (i < n - 1) {
    if (text[i] === "/" && text[i + 1] === "*") {
      let end = text.indexOf("*/", i + 2)
      end = end === -1 ? n : end + 2
      for (let j = i; j < end; j++) {
        if (out[j] !== "\n") out[j] = " "
      }
      i = end
    } else if (i < n - 3 && text.slice(i, i + 4) === "<!--") {
      let end = text.indexOf("-->", i + 4)
      end = end === -1 ? n : end + 3
      for (let j = i; j < end; j++) {
        if (out[j] !== "\n") out[j] = " "
      }
      i = end
    } else {
      i += 1
    }
  }
  return out.join("")
}

/**
 * Characters that, as the last significant token before a `/`, mean the `/`
 * starts a regex literal (not a division operator). After an identifier,
 * number, `)`, or `]` a `/` is division, so we leave it as code.
 */
const REGEX_PREV_CHARS = new Set("(,=:[!&|?{};+-*%<>~^")

/**
 * Last non-whitespace char already emitted to `out` (or '' at start).
 * @param {string[]} out
 * @returns {string}
 */
function prevSignificant(out) {
  for (let k = out.length - 1; k >= 0; k--) {
    const c = out[k]
    if (c !== " " && c !== "\t" && c !== "\r" && c !== "\n") return c
  }
  return ""
}

/**
 * Walk the full file as a single character stream with a state machine.
 * Replace every comment character with a space, preserving line breaks.
 * Strings are preserved as-is.
 *
 * States: code | line_comment | str_dq | str_sq | str_bt | regex
 * (Block comments are handled in stripJsdocBlocks before this.)
 *
 * Regex literals (`/.../flags`) are tracked so that quote characters inside
 * them (e.g. `/"message":"..."/`) don't open a phantom string state and
 * swallow the Hebrew comments that follow.
 *
 * @param {string} input
 * @returns {string}
 */
function stripAllComments(input) {
  const text = stripJsdocBlocks(input)
  /** @type {string[]} */
  const out = []
  /** @type {"code"|"line_comment"|"str_dq"|"str_sq"|"str_bt"|"regex"} */
  let state = "code"
  let inCharClass = false // inside [...] within a regex
  let i = 0
  const n = text.length
  while (i < n) {
    const ch = text[i]
    const nxt = i + 1 < n ? text[i + 1] : ""
    if (state === "code") {
      if (ch === "/" && nxt === "/") {
        out.push("  ")
        state = "line_comment"
        i += 2
        continue
      }
      if (ch === "/" && REGEX_PREV_CHARS.has(prevSignificant(out))) {
        state = "regex"
        inCharClass = false
        out.push(ch)
        i += 1
        continue
      }
      if (ch === '"') state = "str_dq"
      else if (ch === "'") state = "str_sq"
      else if (ch === "`") state = "str_bt"
      out.push(ch)
      i += 1
      continue
    }
    if (state === "line_comment") {
      if (ch === "\n") {
        out.push("\n")
        state = "code"
      } else {
        out.push(" ")
      }
      i += 1
      continue
    }
    if (state === "regex") {
      if (ch === "\\" && i + 1 < n) {
        out.push(ch)
        out.push(text[i + 1])
        i += 2
        continue
      }
      if (ch === "[") inCharClass = true
      else if (ch === "]") inCharClass = false
      else if (ch === "/" && !inCharClass) state = "code"
      else if (ch === "\n") state = "code" // unterminated regex — bail to code
      out.push(ch)
      i += 1
      continue
    }
    // In a string literal: handle escapes, preserve text.
    const quote = state === "str_dq" ? '"' : state === "str_sq" ? "'" : "`"
    if (ch === "\\" && i + 1 < n) {
      out.push(ch)
      out.push(text[i + 1])
      i += 2
      continue
    }
    if (ch === quote) {
      out.push(ch)
      state = "code"
      i += 1
      continue
    }
    out.push(ch)
    i += 1
  }
  return out.join("")
}

/**
 * Return list of [lineNo, originalLine] where Hebrew appears outside comments.
 * @param {string} absPath
 * @returns {Array<[number, string]>}
 */
function scanFile(absPath) {
  let text
  try {
    text = readFileSync(absPath, "utf-8")
  } catch {
    return []
  }
  const codeOnly = stripAllComments(text)
  /** @type {Array<[number, string]>} */
  const violations = []
  const originalLines = text.split("\n")
  const codeLines = codeOnly.split("\n")
  const count = Math.min(originalLines.length, codeLines.length)
  for (let idx = 0; idx < count; idx++) {
    if (HEBREW_RE.test(codeLines[idx])) {
      violations.push([idx + 1, originalLines[idx].replace(/\s+$/, "")])
    }
  }
  return violations
}

/**
 * Recursively collect files with the given extensions under `dir`.
 * @param {string} dir
 * @param {string[]} exts
 * @returns {string[]}
 */
function walk(dir, exts) {
  /** @type {string[]} */
  const found = []
  let entries
  try {
    entries = readdirSync(dir, { withFileTypes: true })
  } catch {
    return found
  }
  for (const entry of entries) {
    const full = join(dir, entry.name)
    if (entry.isDirectory()) {
      found.push(...walk(full, exts))
    } else if (entry.isFile() && exts.some((e) => entry.name.endsWith(e))) {
      found.push(full)
    }
  }
  return found
}

function main() {
  let totalViolations = 0
  /** @type {Array<[string, Array<[number, string]>]>} */
  const badFiles = []

  for (const d of SCAN_DIRS) {
    const root = join(REPO_ROOT, d)
    try {
      if (!statSync(root).isDirectory()) continue
    } catch {
      continue
    }
    for (const path of walk(root, [".ts", ".svelte"])) {
      if (isAllowed(path)) continue
      const v = scanFile(path)
      if (v.length > 0) {
        badFiles.push([path, v])
        totalViolations += v.length
      }
    }
  }

  if (badFiles.length > 0) {
    process.stderr.write("Hebrew strings found in code (must use t('key') instead):\n\n")
    for (const [path, violations] of badFiles) {
      const rel = relative(REPO_ROOT, path).split("\\").join("/")
      process.stderr.write(`  ${rel}\n`)
      for (const [lineNo, line] of violations) {
        process.stderr.write(`    ${lineNo}: ${line}\n`)
      }
      process.stderr.write("\n")
    }
    process.stderr.write(`✗ ${badFiles.length} file(s), ${totalViolations} occurrence(s).\n\n`)
    process.stderr.write("Fix: add the string to packages/core/src/i18n/catalogs/{he,en}.ts,\n")
    process.stderr.write("  add the key to packages/core/src/i18n/keys.ts,\n")
    process.stderr.write("  and use t('your.key') in the component.\n")
    return 1
  }

  process.stdout.write("✓ No hardcoded Hebrew in code.\n")
  return 0
}

// Only run + exit when invoked directly (node lint-no-hebrew-in-code.mjs),
// so test files can import the helpers below without triggering a process exit.
if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  process.exit(main())
}

// Exported for tests.
export { stripAllComments, stripJsdocBlocks, scanFile, isAllowed, HEBREW_RE, main }
